internal/ledgerutil/identifytxs/identifytxs.go

```diff
@@ -0,0 +1,484 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package identifytxs
+
+import (
+	"encoding/hex"
+	"fmt"
+	"math"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger/fabric-protos-go/common"
+	"github.com/hyperledger/fabric/common/ledger/blkstorage"
+	"github.com/hyperledger/fabric/common/metrics/disabled"
+	"github.com/hyperledger/fabric/core/ledger/kvledger"
+	"github.com/hyperledger/fabric/core/ledger/kvledger/txmgmt/rwsetutil"
+	"github.com/hyperledger/fabric/internal/fileutil"
+	"github.com/hyperledger/fabric/internal/ledgerutil/jsonrw"
+	"github.com/hyperledger/fabric/internal/ledgerutil/models"
+	"github.com/hyperledger/fabric/protoutil"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	ledgersDataDirName = "ledgersData"
+	nsJoiner           = "$$"
+	pvtDataPrefix      = "p"
+	hashDataPrefix     = "h"
+)
+
+// IdentifyTxs - IdentifyTxs identifies all transactions related to a list of records (namespace / key pairs)
+// Tool was originally created to map list of divergent records from ledgerutil compare output to their respective transactions
+// to identify list of transactions that have potentially caused divergent state
+// Returns two block numbers that represent the starting and ending blocks of the transaction search range
+func IdentifyTxs(recPath string, fsPath string, outputDirLoc string) (uint64, uint64, error) {
+	// Read diffRecord list from json
+	var records *models.DiffRecordList
+	records, err := jsonrw.LoadRecords(recPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	if len(records.DiffRecords) == 0 {
+		return 0, 0, errors.Errorf("no records were read. JSON record list is either empty or not properly formatted. Aborting identifytxs")
+	}
+
+	// Output directory creation
+	outputDirName := fmt.Sprintf("%s_identified_transactions", records.Ledgerid)
+	outputDirPath := filepath.Join(outputDirLoc, outputDirName)
+
+	empty, err := fileutil.CreateDirIfMissing(outputDirPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	if !empty {
+		return 0, 0, errors.Errorf("%s already exists in %s. Choose a different location or remove the existing results. Aborting identifytxs", outputDirName, outputDirLoc)
+	}
+
+	// Get recordsMap from json
+	inputKeyMapWrapper, err := generateRecordsMap(records.DiffRecords, outputDirPath)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// Set up block store and block iterator in preparation for traversal
+	blockStoreProvider, err := getBlockStoreProvider(fsPath)
+	if err != nil {
+		return 0, 0, err
+	}
+	defer blockStoreProvider.Close()
+
+	blockStore, err := blockStoreProvider.Open(records.Ledgerid)
```

Comment:
Before you Open() you should call blockStoreProvider.Exists() to make sure it exists. Otherwise if user indicates a non-existing channel in the JSON the Open() call will actually create it in the ledger.